src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java

```diff
@@ -363,22 +362,46 @@ else if (abbrStarted) {
     /**
      * Get Abbreviation if it is illegal.
      * @param str name
-     * @param beginIndex begin index
-     * @param endIndex end index
-     * @return true is abbreviation is bigger that required and not in ignore list
+     * @param beginIndex begin index, inclusive
+     * @param endIndex end index, inclusive
+     * @return the abbreviation if bigger than required and not in ignore list
```

Comment:
I thought `@return true if ....` sounded like a method that returns a `boolean`.

What about something like `@return the abbreviation if it is bigger than required and not in the ignore list, otherwise @{code null}`?

I am happy to change this back, though, if you'd prefer!

---

A call to `getAbbreviationIfIllegal(str, n, n)` for a single character abbreviation would return `null` as long as `allowedAbbreviationLength` is not a negative number.

But thinking about `getAbbreviationIfIllegal("Name", 0, 0)`, I guess the edge cases where this _shouldn't_ return `null` would be if the user wanted to enforce "absolutely no capital letters in names" (i.e. setting `allowedAbbreviationLength` to `-1`?), or possibly something like "the capital letter `N` cannot be used in a name"?

Are those valid use cases? If so, I think I could re-work the code a little with that in mind... I guess my initial feeling is that a single-character abbreviation isn't really an abbreviation / passing `0, 0` could maybe be considered invalid. What do you think?

Taking a step back, the existing code called this method with different `endIndex` values depending on whether the abbreviation was at the end of the string or not. My goal was to simplify that by making `endIndex` always the index of the last character in the abbreviation.

Reply:
> `I thought @return true if .... sounded like a method that returns a boolean.`

Ooops my bad. I too just thought it was a boolean return and forgot the real return type.
You can ignore that part of my comment.

> `What about something like @return the abbreviation if it is bigger than required and not in the ignore list, otherwise @{code null}`

This sounds fine to me.

> 0, 0

To me allowing -1 could be valid since we don't prevent it or give an error for it.

My original train of thought was maybe the `inclusive` of this parameter was what is causing the extra code in `getAbbreviation`, specifically the special case of `endIndex == str.length() - 1`.

I want to look at the original code some more based on my other comment.

Reply:
Sounds good. I was thinking about this some more:

In the original code, `getAbbreviationIfIllegal(String str, int beginIndex, int endIndex)` is called with an [exclusive `endIndex` of `str.length()`](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java#L357) when a sequence of upper-case characters stretches to the end of the string, and with an [inclusive `endIndex` of `index - 1`](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java#L346) otherwise (where `index` is the index of a terminating non-upper-case character).

I think this is the source of the bug:

When `getAbbreviationIfIllegal` checks the length of an abbreviation against `allowedAbbreviationLength`, in the original code it considers an abbreviation at the end of a string to be one character longer than it does otherwise.

This makes it falsely flag names like `newCustomerID` as invalid, saying it 'must contain no more than ''2'' consecutive capital letters.'

Reply:
With this PR `getAbbreviationIfIllegal(String str, int beginIndex, int endIndex)` is always called with inclusive indexes, such that `str.substring(beginIndex, endIndex + 1)` is the sequence of upper-case characters.

This makes it so `abbrLength` is computed the same regardless of where in a string an abbreviation occurs.

If `abbrLength > allowedAbbreviationLength`, the abbreviation is illegal unless it is included in the set of `allowedAbbreviations`. (This length check does what we want even when `beginIndex == endIndex`, if `allowedAbbreviationLength == -1`.)

Reply:
@esilkensen 
> > `What about something like @return the abbreviation if it is bigger than required and not in the ignore list, otherwise @{code null}`

> This sounds fine to me.

Please make this change. I am fine with everything else discussed here.